### PR TITLE
[KT-20] [KT-53] Add max_score for separate long problems & add correct/wrong/empty scores for short problems

### PR DESCRIPTION
**Pull request by nugriz**
I made this draft pull request because there is still a lot of unnecesarily comment code and i haven't write for minitest yet but i guess for the main feature is already good

### CODE REVIEW

File: db/migrate/20210203093153_add_correct_to_short_problems.rb [hunk 1]
```diff
@@ -0,0 +1,5 @@
+class AddCorrectToShortProblems < ActiveRecord::Migration
+  def change
+    add_column :short_problems, :correct, :integer, default: 1
```
Comment (donjar) [inline db/migrate/20210203093153_add_correct_to_short_problems.rb]: Can you change these to `correct_score` `wrong_score` `empty_score`? I think these longer names make them much clearer.

File: app/views/short_problems/_manage.html.haml [hunk 2]
```diff
@@ -44,6 +50,18 @@
       = f.label :answer, class: 'col-sm-2'
       .col-sm-4
         = f.text_field :answer, class: 'form-control', autocomplete: 'off'
+    .form-group
```
Comment (donjar) [inline app/views/short_problems/_manage.html.haml:53]: Can you make these input fields default to 1, 0, 0 respectively?

Same goes with the form on short_problems#edit.

File: app/views/long_submissions/mark.html.haml [hunk 3]
```diff
@@ -31,7 +31,7 @@
               %td= tm && tm.mark
               %td= tm && tm.tags
             %td= text_field_tag "marking[#{s.id}][score]", '',
-              placeholder: LongSubmission::SCORE_HASH[s.score],
+              placeholder: s.score == nil ? "-" : s.score.to_s,
```
Comment (donjar) [inline app/views/long_submissions/mark.html.haml]:
```suggestion
              placeholder: s.score.nil? ? "-" : s.score.to_s,
```

File: app/views/contests/summary.html.haml [hunk 4]
```diff
@@ -53,6 +53,6 @@
           %td= lp.problem_no
           = latex_and_markdown :td, lp.statement
           %td= ls.where(score: nil).count
-          - (LongProblem::MAX_MARK + 1).times do |i|
+          - (LongProblem.find(lp.id).max_score + 1).times do |i|
```
Comment (donjar) [inline app/views/contests/summary.html.haml]:
```suggestion
          - (lp.max_score + 1).times do |i|
```

`lp` is already a `LongProblem` right?

File: db/migrate/20210122183013_add_max_score_to_long_problems.rb [hunk 5]
```diff
@@ -0,0 +1,5 @@
+class AddMaxScoreToLongProblems < ActiveRecord::Migration
+  def change
+    add_column :long_problems, :max_score, :integer
```
Comment (donjar) [inline db/migrate/20210122183013_add_max_score_to_long_problems.rb]: Make this default 7?

File: app/views/contests/summary.html.haml [hunk 6]
```diff
@@ -43,7 +43,7 @@
   %h2 Uraian
   %table.table.table-condensed.table-striped.table-responsive
     %thead
-      = create_data_row(['No.', 'Soal', '-'] + (0..LongProblem::MAX_MARK).to_a + ['Rata-rata'],
+      = create_data_row(['No.', 'Soal', '-'] + (0..LongProblem.where(contest_id: params[:id]).maximum("max_score")).to_a + ['Rata-rata'],
```
Comment (donjar) [inline app/views/contests/summary.html.haml]: Can you extract this to a function in the `Contest` model?

e.g.
```ruby
def long_problem_max_score
  long_problems.maximum("max_score")
end
```

File: app/models/concerns/contest_attributes.rb [hunk 7]
```diff
@@ -27,10 +27,10 @@ def currently_in_contest?
     started? && !ended?
   end
 
-  def max_score(sp_count = short_problems.count, lp_count = long_problems.count)
+  def max_score(sp_count = short_problems.count, lp_count = long_problems.count, id_contest)
```
Comment (donjar) [inline app/models/concerns/contest_attributes.rb]: Why does this function take in the `id_contest`?

This concern is loaded in the `Contest` model. It will be a method in that model. You can access the id of that contest directly with the `id` attribute.

File: app/models/concerns/contest_attributes.rb [hunk 8]
```diff
@@ -39,8 +39,8 @@ def results
 
   # This method generates an array containing the number of people getting
   # a certain total score, excluding veterans.
-  def array_of_scores
-    res = Array.new(max_score + 1).fill(0)
+  def array_of_scores(id)
```
Comment (donjar) [inline app/models/concerns/contest_attributes.rb]: Ditto, this does not need an `id` argument.

File: app/models/concerns/contest_attributes.rb [hunk 9]
```diff
@@ -27,10 +27,10 @@ def currently_in_contest?
     started? && !ended?
   end
 
-  def max_score(sp_count = short_problems.count, lp_count = long_problems.count)
+  def max_score(sp_count = short_problems.count, lp_count = long_problems.count, id_contest)
     sp_count ||= short_problems.count
```
Comment (donjar) [inline app/models/concerns/contest_attributes.rb]: These are not needed anymore, right?

File: app/models/concerns/user_contest_scope.rb [hunk 10]
```diff
@@ -17,7 +17,7 @@ module UserContestScope
         end
         .group(:id)
         .select('user_contests.id as id, sum(case when ' \
-        'short_submissions.answer = short_problems.answer then 1 else 0 end) ' \
+        'short_submissions.answer = short_problems.answer then short_problems.correct when short_submissions.answer = null then short_problems.empty else short_problems.wrong end) ' \
```
Comment (donjar) [inline app/models/concerns/user_contest_scope.rb]: You might also need to check for empty string (`""`).

File: app/models/long_submission.rb [hunk 11]
```diff
@@ -40,11 +40,23 @@ class LongSubmission < ActiveRecord::Base
 
   validates :score,
             numericality: { allow_nil: true, greater_than_or_equal_to: 0 }
+  def get_score_text
```
Comment (donjar) [inline app/models/long_submission.rb:51]: Duplicate methods?

File: app/models/long_submission.rb [hunk 12]
```diff
@@ -40,11 +40,23 @@ class LongSubmission < ActiveRecord::Base
...
+    #scorehash = [*0..LongProblem.maximum("max_score")].each_with_object({}) do |item, memo|
+    #  memo[item] = item.to_s
+    #end
+    #scorehash[nil] = '-'
+    #return scorehash
+    return "-" if score.nil?
+    return score.to_s
```
Comment (donjar) [inline app/models/long_submission.rb]: You can omit `return` on the last statement on Ruby functions. Ruby functions always return the value of the last statement in the function body.

File: app/models/long_submission.rb [hunk 13]
```diff
@@ -40,11 +40,23 @@ class LongSubmission < ActiveRecord::Base
...
+  end
 
-  SCORE_HASH = [*0..LongProblem::MAX_MARK].each_with_object({}) do |item, memo|
-    memo[item] = item.to_s
+  def get_text_score(x)
+    if x == "-"
+        return nil
+    else
+        return x.to_i
```
Comment (donjar) [inline app/models/long_submission.rb]: See above

File: app/models/user.rb [hunk 14]
```diff
@@ -82,7 +82,7 @@ class User < ActiveRecord::Base
 
   def before_add_method(role)
     return if !Role.admins.include?(role.name) || has_role?(:panitia)
-    raise 'User is not panitia!'
+    raise 'User is not panitia! Run `User.find(your_id).add_role :panitia` first.'
```
Comment (donjar) [inline app/models/user.rb]: Why is this here?

Might it be that you have not rebased from master? Basically, you did two commits in your past PR ("add solution if you're not panitia" and "Add suggested action to be panitia before being admin"), but they are squashed into one commit when the previous PR is merged into master.

So right now the git tree is like this:
```
X - Y (master)
|
Y1 - Y2 - Z1 - Z2 - Z3 - Z4 (your branch)
```

Previously, when working on the panitia stuff you added the commits Y1 and Y2. They are combined into one commit Y in master. However, looking at the commit history of this branch (https://github.com/ktomatematika/kontes-terbuka/pull/126/commits) it seems like you still have the two commits Y1 and Y2.

Now we want the git tree to be like this
```
X - Y (master)
    |
    Z1 - Z2 - Z3 - Z4 (your branch)
```

To achieve this, we want to rebase from master. Rebase simply means, we want to base your branch into a different commit: from commit X as base, to commit Y as base. However, we also want to remove these two commits Y1 and Y2.

To do this, you want to use interactive rebase. You can do so from your branch, by running `git rebase -i master` (where -i stands for interactive). When you run this command, you will see a text editor pop up, with a line describing each commit, preceded by the word "pick". You can either replace the word "pick" by "drop" on these two commits Y1 and Y2, or you can just delete these two lines. By doing so, you will be left with just the 4 commits Z1 to Z4.

Do ping me if you feel that this is unclear :)

File: app/models/user_contest.rb [hunk 15]
```diff
@@ -75,7 +75,7 @@ def contest_points
     ls_points = long_submissions.reduce(0) do |memo, item|
       case item.score
       when nil then memo
-      when 0..(LongProblem::MAX_MARK - 1) then memo + 1 # +1 if score not nil
+      when 0..((LongProblem.find(item.long_problem_id).max_score) - 1) then memo + 1 # +1 if score not nil
```
Comment (donjar) [inline app/models/user_contest.rb]:
```suggestion
      when 0..(item.long_problem.max_score) - 1) then memo + 1 # +1 if score not nil
```

File: app/helpers/contests_helper.rb [hunk 16]
```diff
@@ -57,9 +57,17 @@ def percentage(num)
...
-    LongSubmission::SCORE_HASH[user_contest.__send__(
-      'problem_no_' + long_problem.id.to_s
-    )]
+    #hash_score = LongSubmission.find_by(long_submission_id: long_problem.id).score_hash
+    #hash_score[user_contest.__send__(
+    #  'problem_no_' + long_problem.id.to_s
+    #)]
+    #x = LongSubmission.find_by(user_contest_id: user_contest.id, long_problem_id: long_problem.id)
+    #x.get_score_text
+    if user_contest.__send__('problem_no_' + long_problem.id.to_s) == nil
```
Comment (donjar) [inline app/helpers/contests_helper.rb]: Can't you use `get_score_hash` here?

File: app/helpers/contests_helper.rb [hunk 17]
```diff
@@ -88,7 +96,7 @@ def average_mark(long_submissions)
 
   # Helper for contests#_own_results.
   def score_out_of_total(lp)
-    score(@user_contest, lp).to_s + '/' + LongProblem::MAX_MARK.to_s + ' poin'
+    score(@user_contest, lp).to_s + '/' + LongProblem.find(lp.id).max_score.to_s + ' poin'
```
Comment (donjar) [inline app/helpers/contests_helper.rb]:
```suggestion
    score(@user_contest, lp).to_s + '/' + lp.max_score.to_s + ' poin'
```

**Review by donjar — COMMENTED** [review]
Thanks a lot for this @nugriz! In general they look good. Great work, kudos to you!

I left some suggestions. They might seem a lot but all of them should be quite minor.

Do ping me again if you have added tests :)

Also, do note our branch naming convention of using dashes with all small letters (so should be `release-latest-scoring-feature`). This is alright, we can leave it so, but do take note of it next time. In addition, please change your PR title to put in the JIRA ticket titles (i.e. KT-20 and KT-53; your PR title should be prefixed with "[KT-20] [KT-53] Release latest scoring feature").

File: db/migrate/20210203093153_add_correct_to_short_problems.rb (hunk 1 above)
Comment (nugriz) [reply db/migrate/20210203093153_add_correct_to_short_problems.rb]: okay

File: app/views/short_problems/_manage.html.haml (hunk 2 above)
Comment (nugriz) [reply app/views/short_problems/_manage.html.haml:53]: In my local machine, it is already there
image

File: db/migrate/20210122183013_add_max_score_to_long_problems.rb (hunk 5 above)
Comment (nugriz) [reply db/migrate/20210122183013_add_max_score_to_long_problems.rb]: okay

File: app/views/contests/summary.html.haml (hunk 6 above)
Comment (nugriz) [reply app/views/contests/summary.html.haml]: okay

File: app/models/concerns/user_contest_scope.rb (hunk 10 above)
Comment (nugriz) [reply app/models/concerns/user_contest_scope.rb]: what the different between "" and null

File: app/helpers/contests_helper.rb (hunk 16 above)
Comment (nugriz) [reply app/helpers/contests_helper.rb]: do you mean get_score_text?

File: app/views/short_problems/_manage.html.haml (hunk 2 above)
Comment (donjar) [reply app/views/short_problems/_manage.html.haml:53]: I see. That's probably from the database defaults, so I guess it should be all good

File: app/models/concerns/user_contest_scope.rb (hunk 10 above)
Comment (donjar) [reply app/models/concerns/user_contest_scope.rb]: I'm actually not sure. I suspect that the client might be able to send an empty string to our servers and it'll be the empty string in the database. We still want to treat them as an empty answer though instead of a wrong one.

File: app/helpers/contests_helper.rb (hunk 16 above)
Comment (donjar) [reply app/helpers/contests_helper.rb]: Oh yeah probably that

File: app/controllers/temporary_markings_controller.rb [hunk 18]
```diff
@@ -16,7 +16,7 @@ def modify_on_long_problem
       mark = val[:mark]
       tags = val[:tags]
 
-      update_hash = { mark: LongSubmission::SCORE_HASH.key(mark), tags: tags }
+      update_hash = { mark: LongSubmission.text_to_score(mark) , tags: tags }
```
Comment (donjar) [inline app/controllers/temporary_markings_controller.rb]:
```suggestion
      update_hash = { mark: LongSubmission.text_to_score(mark), tags: tags }
```

File: app/models/concerns/user_contest_scope.rb [hunk 19]
```diff
@@ -17,7 +17,7 @@ module UserContestScope
         end
         .group(:id)
         .select('user_contests.id as id, sum(case when ' \
-        'short_submissions.answer = short_problems.answer then 1 else 0 end) ' \
+        "short_submissions.answer = short_problems.answer then short_problems.correct_answer when short_submissions.answer = null then short_problems.empty_score when short_submissions.answer = '' then short_problems.empty_score else short_problems.wrong_score end) " \
```
Comment (donjar) [inline app/models/concerns/user_contest_scope.rb]:
```suggestion
        "short_submissions.answer = short_problems.answer then short_problems.correct_answer when short_submissions.answer = null then short_problems.empty_score when short_submissions.answer = '' then short_problems.empty_score else short_problems.wrong_score end)" \
```

File: test/models/concerns/contest_attributes_test.rb [hunk 20]
```diff
@@ -78,16 +78,16 @@ class ContestAttributesTest < ActiveSupport::TestCase
       create(:short_problem, contest: c1, problem_no: i + 1)
     end
     7.times.each do |i|
-      create(:long_problem, contest: c1, problem_no: i + 1)
+      create(:long_problem, contest: c1, problem_no: i + 1, max_score: 10)
     end
     5.times.each do |i|
       create(:short_problem, contest: c2, problem_no: i + 1)
```
Comment (donjar) [inline test/models/concerns/contest_attributes_test.rb]: Can you also put `correct_score: 4` here?

The idea is that even though it is already specified in `factories.rb` we want developers who read this test to understand this test easily without looking at any other files. Readability is always our main goal.

File: app/views/long_submissions/mark.html.haml [hunk 21]
```diff
@@ -31,7 +31,7 @@
               %td= tm && tm.mark
               %td= tm && tm.tags
             %td= text_field_tag "marking[#{s.id}][score]", '',
-              placeholder: LongSubmission::SCORE_HASH[s.score],
+              placeholder: s.score.nil? ? "-" : s.score.to_s,
```
Comment (donjar) [inline app/views/long_submissions/mark.html.haml]: Use get_score_text?

**Review by donjar — COMMENTED** [review]
Please add tests for your new features as well :)

File: test/models/user_contest_test.rb [hunk 22]
```diff
@@ -69,7 +69,7 @@ class UserContestTest < ActiveSupport::TestCase
                               gold_cutoff: 7)
     ucs = c.user_contests
     sp = c.short_problems.take
-
+    c.long_problems.first.update(max_score: 7)
```
Comment (donjar) [inline test/models/user_contest_test.rb:72]: Why is this needed?

File: test/models/user_contest_test.rb [hunk 23]
```diff
@@ -95,6 +95,24 @@ class UserContestTest < ActiveSupport::TestCase
...
+    sp4 = c.short_problems.fourth
+    sp3.update(answer: 1)
+    none = ucs.first
+    none.short_submissions.first.update(answer: sp1.answer)
+    none.short_submissions.second.update(answer: nil)
+    none.short_submissions.third.update(answer: '')
+    none.short_submissions.fourth.update(answer: sp4.answer.to_i + 1)
+
+    pucs = ucs.processed
+    assert_equal pucs.find(none.id).short_mark, 1
```
Comment (donjar) [inline test/models/user_contest_test.rb]: Are you not testing for other cases?

File: app/models/long_submission.rb [hunk 24]
```diff
@@ -40,11 +40,15 @@ class LongSubmission < ActiveRecord::Base
...
   validates :score,
             numericality: { allow_nil: true, greater_than_or_equal_to: 0 }
+  def get_score_text
+    return "-" if score.nil?
+    score.to_s
+  end
 
-  SCORE_HASH = [*0..LongProblem::MAX_MARK].each_with_object({}) do |item, memo|
-    memo[item] = item.to_s
+  def text_to_score(x)
```
Comment (donjar) [inline app/models/long_submission.rb]: Isn't this supposed to be a class method?

File: app/models/user.rb (hunk 14 above)
Comment (nugriz) [reply app/models/user.rb]: I found problem :
fatal: invalid upstream 'master'
when i was googling, solution from stacoverflow is :
git rebase -i origin/master
I'm not sure

File: app/models/long_submission.rb (hunk 24 above)
Comment (nugriz) [reply app/models/long_submission.rb]: It was, but when i realized that the long_submission score must not exceed max_score, i need the max_score so i bind it then become object method, i guess

File: test/models/user_contest_test.rb (hunk 23 above)
Comment (nugriz) [reply test/models/user_contest_test.rb]: Can you give me some example?

Comment (donjar) [reply test/models/user_contest_test.rb]: Oh wait I see what you're doing here. So the thing about tests is that you want to try to isolate every feature in separate tests. This is so that we can detect if there is one specific feature that is wrong. In this particular instance it is advisable to have one test for correct answers; one test for wrong answers; and one test for empty answers.

File: app/models/long_submission.rb (hunk 24 above)
Comment (donjar) [reply app/models/long_submission.rb]: I see. Looking at the usages of this method I would think that checking for exceeding max_score might be misleading. Imagine if you are marking a submission and you accidentally put score 8 when the max score is 7. Now it will be stored as nil and the marker will not be alerted for this. I think it's much better to have validations for the score of the long submission to not exceed the long problem's max score instead.

File: app/models/user.rb (hunk 14 above)
Comment (donjar) [reply app/models/user.rb]: Oh sorry our main branch is called `production`. So it should be `git rebase -i production` instead.

**Review by donjar — APPROVED** [review]
